File: backend/schemas/jumper.py
from pydantic import BaseModel, ConfigDict
from typing import List, Optional
from .document import TraitResponse

# ==========================================
# BUILD SCHEMAS
# ==========================================
class BuildBase(BaseModel):
    jumper_id: int
    document_id: int

class BuildCreate(BuildBase):
    trait_ids: List[int] = []
    age: Optional[int] = None
    gender: Optional[str] = None

class BuildUpdate(BaseModel):
    trait_ids: List[int] = []
    age: Optional[int] = None
    gender: Optional[str] = None

class BuildResponse(BuildBase):
    id: int
    remaining_cp: int
    traits: List[TraitResponse] = []
    age: int
    gender: str
    model_config = ConfigDict(from_attributes=True)

# ==========================================
# JUMPER SCHEMAS
# ==========================================
class JumperBase(BaseModel):
    name: str
    age: int
    gender: str

class JumperCreate(JumperBase):
    pass

class JumperResponse(JumperBase):
    id: int
    builds: List[BuildResponse] = []
    model_config = ConfigDict(from_attributes=True)

File: frontend/src/types/jumper.ts
import type { Trait } from "../schemas/documentSchema";

export interface BuildStats {
    remainingCp: number;
    catCounts: Record<number, number>;
    traitMap: Map<number, { trait: Trait; catId: number }>;
    lockedTraits: Set<number>;
    finalCosts: Map<number, number>;
    bypassAgeRoll: boolean;
    bypassGenderLock: boolean;
}

export interface Jumper {
    id: number;
    name: string;
    age: number;
    gender: string;
    builds?: Build[];
}

export interface Build {
    id: number;
    jumper_id: number;
    document_id: number;
    remaining_cp: number;
    age: number;
    gender: string;
    traits: Trait[];
}